subsys/bluetooth/host/keys.c

```diff
@@ -279,6 +279,22 @@ void bt_keys_add_type(struct bt_keys *keys, int type)
 	keys->keys |= type;
 }
 
+void bt_keys_clear_other_keys_with_identity(uint8_t id, const bt_addr_le_t *id_addr,
+					    const struct bt_keys *keep_these_keys)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(key_pool); i++) {
+		if (&key_pool[i] == keep_these_keys) {
+			/* Skip the keep_this_key */
+			continue;
+		}
+
+		if ((key_pool[i].id == id) && (!bt_addr_le_cmp(&key_pool[i].addr, id_addr))) {
+			/* Clear this key, because it has the same id and identity address. */
```

Comment:
I think this might bring bt_dev.le.rl_entries in a bad state since the entry cleared here is an entry for the same device, buit from the previous entry. We are trying to keep the number of keys the same since smp_ident_addr_info is overwriting the RPA (used as the address in key_pool for the LTK before the IRK and IA is known) with the IA, but it should have overwritten the previous entry instead (which is now done by clearing the old entry and updating the new entry).

Do you agree?

Reply:
You are right @sjanc ! Calling bt_id_del in bt_keys_clear_other_keys_with_identity() and leaving bt_id_add() unchanged works. I tested it on my setup, and it can handle an updated IRK upon rebonding.